metaflow/plugins/deployer.py

```diff
@@ -0,0 +1,192 @@
+from metaflow.runner.utils import read_from_file_when_ready
+
+
+def handle_timeout(tfp_runner_attribute, command_obj: CommandManager):
+    try:
+        content = read_from_file_when_ready(tfp_runner_attribute.name, timeout=10)
+        return content
+    except TimeoutError as e:
+        stdout_log = open(command_obj.log_files["stdout"]).read()
+        stderr_log = open(command_obj.log_files["stderr"]).read()
+        command = " ".join(command_obj.command)
+        error_message = "Error executing: '%s':\n" % command
+        if stdout_log.strip():
+            error_message += "\nStdout:\n%s\n" % stdout_log
+        if stderr_log.strip():
+            error_message += "\nStderr:\n%s\n" % stderr_log
+        raise RuntimeError(error_message) from e
+
+
+def get_lower_level_group(api, top_level_kwargs: Dict, _type: str, name: Optional[str]):
+    if _type is None:
+        raise ValueError(
+            "Deployer doesn't have a 'type' to target. Please use a sub-class of Deployer."
+        )
+    if name is None:
+        return getattr(api(**top_level_kwargs), _type)()
+    return getattr(api(**top_level_kwargs), _type)(name=name)
+
+
+class TriggeredRun(object):
+    def __init__(
+        self,
+        deployer: "Deployer",
+        content: str,
+    ):
+        self.deployer = deployer
+        content_json = json.loads(content)
+        self.metadata_for_flow = content_json.get("metadata")
+        self.pathspec = content_json.get("pathspec")
+        self.deployer.name = content_json.get("name")
+
+    @property
+    def name(self):
+        return self.deployer.name
+
+    @property
+    def status(self):
+        raise NotImplementedError
+
+    @property
+    def run(self):
+        clear_and_set_os_environ(self.deployer.old_env)
+        metadata(self.metadata_for_flow)
+        try:
+            return Run(self.pathspec, _namespace_check=False)
+        except MetaflowNotFound:
+            return None
+
+    def terminate(self, **kwargs):
+        _, run_id = self.pathspec.split("/")
+        command = get_lower_level_group(
+            self.deployer.api,
+            self.deployer.top_level_kwargs,
+            self.deployer.type,
+            self.deployer.name,
+        ).terminate(run_id=run_id, **kwargs)
+
+        pid = self.deployer.spm.run_command(
+            [sys.executable, *command],
+            env=self.deployer.env_vars,
+            cwd=self.deployer.cwd,
+            show_output=self.deployer.show_output,
+        )
+
+        command_obj = self.deployer.spm.get(pid)
+        return command_obj.process.returncode == 0
+
+
+class DeployedFlow(object):
+    def __init__(
+        self,
+        deployer: "Deployer",
+    ):
+        self.deployer = deployer
+
+    @property
+    def production_token(self):
+        raise NotImplementedError
+
+    def trigger(self, **kwargs) -> "TriggeredRun":
+        return self.deployer.trigger(**kwargs)
+
+
+class Deployer(object):
```

Comment:
I am debating whether we should have a common thing like this or let each CLI define everything individually. Pros and cons on both sides.